### ISSUE TITLE
Android: Downloaded Dictionary is not in list

### ISSUE THREAD

**Issue by haymant** [koreader/koreader#6613]
* KOReader version: koreader-android-arm-v2020.08.1.apk
* Device: Android 10 on Huawei P30Pro

![android](https://user-images.githubusercontent.com/700512/91778755-1432b000-ec26-11ea-875f-0c7f63f35141.jpg)

#### Issue

I can download dictionary (e.g., the GNU english dictioanary), the log shows:
```
09-01 07:36:25.552 17918 17961 W KOReader: command ./tar xf /storage/emulated/0/koreader/data/dict/gcide.tar.gz -C /storage/emulated/0/koreader/data/dict returned 139
```
If I try again, we can see koreader knows the dictionary is downloaded
![install](https://user-images.githubusercontent.com/700512/91778790-2dd3f780-ec26-11ea-8f06-e1f09ead1474.jpg)

However, the dictionary does not appear in list.
![list](https://user-images.githubusercontent.com/700512/91778746-0c730b80-ec26-11ea-8be1-f1eff6a95bb0.jpg)

#### Steps to reproduce

1. Goto Dictionary Download menu, download an english dictionary
2. Back one level up
3. The downloaded dictionary is not in list

**Comment by pazos** [comment]
> 09-01 07:36:25.552 17918 17961 W KOReader: command ./tar xf /storage/emulated/0/koreader/data/dict/gcide.tar.gz -C  returned 139

Maybe we need to extract dicts into the internal cache and copy the extracted files to /storage/emulated/0/koreader/data/dict? I never played with dict downloads on android (other than avoiding ANRs while downloading)

**Comment by Frenzie** [comment]
@pazos Are you saying the dictionary itself still works, just not the built-in download?

**Comment by pazos** [comment]
stardict dictionaries work fine, yeah. Not sure about the download but it seems to fail uncompressing.

**Comment by haymant** [comment]
Android 10 didn't untar at all (i.e., unlike Android 8.1 device I tried, which got incomplete directory), I untared them and these dictionary got listed but then stardict returned "no definition found"

**Comment by tryfon** [comment]
I have almost exactly the same issue. In my case the dictionary  was extracted successfully but is not listed. Huawei Media Pad T5, Android 8.0, koreader v2020.08.1
Please prioritize this or provide a workaround. Embedded dictionary is a quite essential feature.

**Comment by NiLuJe** [comment]
The workaround is you unpacking the download manually ;).

There's not much we can do about Android actively working against us...

**Comment by tryfon** [comment]
It would be appreciated if you read my comment more carefully. In my case the compressed file was extracted but still it is not working.

**Comment by Frenzie** [comment]
> provide a workaround.

Please try that "use external dictionary" toggle? ;-)

**Comment by tryfon** [comment]
Tried that already and it doesn't seem to be working.

**Comment by Frenzie** [comment]
Doesn't seem to be working how? First you have to tick it, so an external app will be used instead of the built-in one, and then you have to select the external app of your choice in the menu below. You will need to have one of the supported apps installed of course, but you can add your own too (and submit PRs).

**Comment by tryfon** [comment]
OK, I managed to make it work by installing and configuring the Aard2 external dictionary with a Wordnet dictionary. Still not the seamless experience that the internal feature provides but it is OK. Thank you!

**Comment by NiLuJe** [comment]
I suspect there's still permission issues involved, which is why I stand by the "do it by hand" comment, FWIW ;).

**Comment by pazos** [comment]
Very very broken issue here. Intended in fixing I'm not. Shall we disable it? Shall we just redirect the user to a webpage with dictionaries?

**Comment by NiLuJe** [comment]
@pazos: Sounds good to me. Unpacking stuff manually works for me on P.

**Comment by Frenzie** [comment]
> Shall we just redirect the user to a webpage with dictionaries?

On Android that shouldn't be much of a problem. I threw this together mainly for Kobo and similar platforms after all.

*However*, this functionality works perfectly fine on my Zenfone 2 with Android 6. It's not the case that it's just flat-out broken on all versions of Android, so it probably makes sense only to do so on 8+?

**Comment by pazos** [comment]
> However, this functionality works perfectly fine on my Zenfone 2 with Android 6. It's not the case that it's just flat-out broken on all versions of Android, so it probably makes sense only to do so on 8+?

Duh! I don't know. A proper android fix would be:

1. for each dictionary generate an UUID
2. create a dir in cache with UUID as name
3. download to cache dir
4. execute a callback after uncompressing to cache to move recursively from cache to data dir (and remove leftovers?, if not cache will be free eventually, but that depends on the OS).

That should work on any API but needs some work on the backend to move things recursively. The rest can be achieved easily.

**Comment by Frenzie** [comment]
I meant that simply as a less invasive interim solution. ^_^ `if android goto website` vs `if android >= 8 goto website` is practically the same thing after all.

**Comment by Galunid** [comment]
@tryfon If you're still interested, you can delete the extracted directory, and extract it yourself, to get it working properly.
@pazos Perhaps letting users know they should extract the dictionary themselves should be ok? I tested it on 2 android 8+ devices and it works fine if I remove the extracted files and extract them myself again.

**Comment by haymant** [comment]
agree. dont want to turn on wifi for dictionary

**Comment by pazos** [comment]
> Perhaps letting users know they should extract the dictionary themselves should be ok? I tested it on 2 android 8+ devices and it works fine if I remove the extracted files and extract them myself again.

I'm not even sure what's the root issue here. Maybe tar is trying to preserve file permissions?. Anyways whatever solution is better than disabling the whole thing on android.

**Comment by NiLuJe** [comment]
@pazos: Yeah, that was my hunch. From what I recall, it does extract *some* stuff, but not all, or something.

**Comment by retrue** [comment]
Hi, I was going to open an issue here about this but I used the search feature  :p
I'll tell my experience: I just got an Android phone with Android 10.
I didn't have any dictionary so I downloaded the English Wordnet directory. It then appears listed in installed dictionaries, but when I look for any word it appears a message saying "Entry not found" (it doesn't matter the word searched) and the menu bar on top of the dictionary window appears empty instead of showing the name of the dictionary.
Then I tried and downloaded other two dictionaries: an English/Spanish and the GNU English dictionaries. In this case, these dictionaries don't even appear in the list of installed dictionaries. But if I try to download them again, KOReader asks about overwriting the files.
Looking in Data/Dict they appear as .gz files with a folder each one and one file in each folder. A .ifo file for the Wordnet dictionary and a .idx file in the case of the other two dictionaries.
I guess I'll need to install an app to uncompress gz files...
